### PR TITLE
XRay Guide using OpenTelemetry

### PR DESCRIPTION
**Pull request by sdelamo**
Steps to build the locally: 

```
./gradlew micronautMicroservicesDistributedTracingXrayBuild -x micronautMicroservicesDistributedTracingXrayRunTestScript
```

in `build.gradle` ensure your dependencies are: 

```groovy
    annotationProcessor(platform("io.micronaut.tracing:micronaut-tracing-bom:4.2.0"))
    implementation(platform("io.micronaut.tracing:micronaut-tracing-bom:4.2.0"))
```

to pull 4.2.0 which has the xray tracing in it.

Run `aws configure` to choose a region (us-east-1)

Clone  `https://github.com/aws-observability/aws-otel-collector`

Edit `examples/docker/config-test.yaml` and use `us-east-1`

and run the Aws OpenTelementry Collector
```
docker run --rm -p 4317:4317 -p 55680:55680 -p 8889:8888 -e AWS_REGION=us-east-1 -e AWS_PROFILE=default -v ~/.aws:/root/.aws -v "${PWD}/examples/docker/config-test.yaml":/otel-local-config.yaml --name awscollector public.ecr.aws/aws-observability/aws-otel-collector:latest --config otel-local-config.yaml
```

Run dynamodb locally:

```
docker run -it --rm \
     -p 8000:8000 \
     amazon/dynamodb-local
```

And add:

```
dynamodb-local:
  host: localhost
  port: 8000
```

To both application-dev.yml files

### CODE REVIEW

**Review by timyates — CHANGES_REQUESTED** [review]
This is missing the part about setting up

```
dynamodb-local:
  host: localhost
  port: 8000
```

In `application-dev.yml` for bookinventory and bookcatalogue

And also it needs [`src/docs/common/snippets/common-run-dynamodb-with-docker.adoc`](https://github.com/micronaut-projects/micronaut-guides/pull/996/files#diff-43144f350447b277b18bc1101d3a3b9bc7e9c7f566db1ec95ee3fa6c61559984) from, the dynamodb guide PR injected into to asciidoc (as without a local running dynamodb, `./gradlew run` fails)

**Comment by timyates** [comment]
Updated the issue description now xray support is out in tracing:4.2.0

File: guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+* `docker run --rm -p 4317:4317 -p 55680:55680 -p 8889:8888 -e AWS_REGION=us-east-1 -e AWS_PROFILE=default -v ~/.aws:/root/.aws -v "${PWD}/examples/docker/config-test.yaml":/otel-local-config.yaml --name awscollector public.ecr.aws/aws-observability/aws-otel-collector:latest --config otel-local-config.yaml`
+
+== AWS Command Line Interface
+
+Install the https://aws.amazon.com/cli/[AWS Command Line Interface] and run `aws configure`.
+Set the same region you used in the previous section.
+
+== OpenTelemetry dependencies
+
+=== OpenTelemetry annotations
```
Comment (shingjo) [inline guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc]: There isn't any OT annotations.

**Review by shingjo — APPROVED** [review]

**Comment by sdelamo** [comment]
The classes under `buildSrc` will be removed once 3.6.0 is out

**Comment by sdelamo** [comment]
@kroegerk can you proofread it?

File: guides/micronaut-microservices-distributed-tracing-xray/metadata.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
+{
+  "asciidoctor": "micronaut-microservices-distributed-tracing-xray.adoc",
+  "slug": "micronaut-microservices-distributed-tracing-xray",
+  "title": "Microservices distributed tracing with XRay and Micronaut Framework",
```
Comment (kroegerk) [inline guides/micronaut-microservices-distributed-tracing-xray/metadata.json]:
```suggestion
  "title": "Microservices Distributed Tracing with X-Ray and the Micronaut Framework",
```

File: guides/micronaut-microservices-distributed-tracing-xray/metadata.json [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  "intro": "Use XRay distributed tracing to investigate the behaviour of your Micronaut applications.",
```
Comment (kroegerk) [inline guides/micronaut-microservices-distributed-tracing-xray/metadata.json]:
```suggestion
  "intro": "Use AWS X-Ray distributed tracing to investigate the behavior of your Micronaut applications.",
```

File: guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc [hunk 4]
```diff
@@ -0,0 +1,207 @@
+common:header-top.adoc[]
+
+== Getting Started
+
+In this guide, we are going to use https://opentelemetry.io[OpenTelemetry] to collect distributed
+tracing information of a Micronaut application composed of three microservices and publish it to https://aws.amazon.com/xray/[AWS X-Ray]
+
+____
+AWS X-Ray helps developers analyze and debug production, distributed applications, such as those built using a microservices architecture. With X-Ray, you can understand how your application and its underlying services are performing to identify and troubleshoot the root cause of performance issues and errors. X-Ray provides an end-to-end view of requests as they travel through your application, and shows a map of your application’s underlying components. You can use X-Ray to analyze both applications in development and in production, from simple three-tier applications to complex microservices applications consisting of thousands of services.
```
Comment (kroegerk) [inline guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc]:
```suggestion
AWS X-Ray helps developers analyze and debug distributed applications, such as those built using a microservices architecture. With X-Ray, you can understand how your application and its underlying services are performing to identify and troubleshoot the root cause of performance issues and errors. X-Ray provides an end-to-end view of requests as they travel through your application and displays a map of your application’s underlying components. You can use X-Ray to analyze both applications in development and in production, from simple three-tier applications to complex microservices applications consisting of thousands of services.
```

I know you copied this directly from the AWS site, but the first sentence doesn't make sense as written and the comma I removed does not belong there. :-)

File: guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+common:requirements.adoc[]
+
+common:completesolution.adoc[]
+
+== Writing the application
+
+The application contains three microservices.
+
+* `bookcatalogue` - It returns a list of books. It uses a domain consisting of a book name and ISBN.
```
Comment (kroegerk) [inline guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc]:
```suggestion
* `bookcatalogue` - Returns a list of books. It uses a domain consisting of a book name and an ISBN.
```

File: guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+* `bookinventory` - It exposes an endpoint to check whether a book has sufficient stock to fulfill an order. It uses a domain consisting of a stock level and ISBN.
```
Comment (kroegerk) [inline guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc]:
```suggestion
* `bookinventory` - Exposes an endpoint to check whether a book has sufficient stock to fulfill an order. It uses a domain consisting of a stock level and an ISBN.
```

File: guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+* `bookrecommendation` - It consumes previous services and exposes an endpoint recommending book names in stock.
```
Comment (kroegerk) [inline guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc]:
```suggestion
* `bookrecommendation` - Consumes previous services and exposes an endpoint recommending book names in stock.
```

File: guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+=== OpenTelemetry annotations
+
+To enable OpenTelemetry annotations, every service includes the next annotation processor:
+
+dependency:micronaut-tracing-opentelemetry-annotation[groupId=io.micronaut.tracing,scope=annotationProcessor]
+
+=== Micronaut OpenTelemetry HTTP
+
+To enable creating span objects on the every HTTP server request, client request, server response and client response
```
Comment (kroegerk) [inline guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc]:
```suggestion
To enable creation of span objects on every HTTP server request, client request, server response, and client response,
```

File: guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+An exporter is a component in the OpenTelemetry Collector configured to send data to different systems/back-ends.
+
+Each service adds the OpenTelemetry Protocol exporter dependency.
+
+dependency:opentelemetry-exporter-otlp[groupId=io.opentelemetry,scope=compile]
+
+=== OpenTelemetry Propagator
+
+To use AWS X-Ray propagator each service adds the following dependency:
```
Comment (kroegerk) [inline guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc]:
```suggestion
To use AWS X-Ray propagator, each service adds the following dependency:
```

File: guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+
+The previous configuration excludes health checks from being traced.
+
+===  Using AWS resource detector
+
+Each service includes the following dependency:
+
+dependency:opentelemetry-aws-xray[groupId=io.opentelemetry.contrib,scope=compile]
+
+They allow you to enable https://aws-otel.github.io/docs/getting-started/java-sdk/trace-manual-instr#using-the-aws-resource-detectors[AWS resource detectors for enriching traces
```
Comment (kroegerk) [inline guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc]: It's not clear here what "they" refers to.

```suggestion
These dependencies allow you to enable https://aws-otel.github.io/docs/getting-started/java-sdk/trace-manual-instr#using-the-aws-resource-detectors[AWS resource detectors for enriching traces
```

File: guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+
+Both services use DynamoDB and include the following dependencies:
+
+:dependencies:
+dependency:micronaut-aws-sdk-v2[groupId=io.micronaut.aws,scope=compile]
+dependency:dynamodb[groupId=software.amazon.awssdk]
+:dependencies:
+
+`micronaut-aws-sdk-v2` dependency creates a bean of type `SdkClientBuilder`. To instrument the AWS SDK,
+Micronaut OpenTelemetry registers a tracing interceptor by creating a bean creation listener for the bean of type `SdkClientBuilder`.
```
Comment (kroegerk) [inline guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc]:
```suggestion
Micronaut OpenTelemetry registers a tracing interceptor by creating a bean-creation listener for the bean of type `SdkClientBuilder`.
```

File: guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+
+You can run a cURL command to test the whole application:
+
+[source, bash]
+----
+$ curl http://localhost:8080/books
+[{"name":"Building Microservices"}
+----
+
+You can then navigate to AWS Console and access the X-Ray UI
```
Comment (kroegerk) [inline guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc]:
```suggestion
You can then navigate to AWS Console and access the X-Ray UI.
```

File: guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+In the previous image, you can see that:
+
+- Whenever a Micronaut HTTP client executes a new network request, it creates a new subsegment.
+- Whenever a Micronaut server receives a request, it creates a new segment.
+
+Moreover, you can see the requests to `bookinventory` are made in parallel.
+
+== Next steps
+
+As you have seen in this guide, you get distributing tracing up-and-running fast with the Micronaut framework without any annotations.
```
Comment (kroegerk) [inline guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc]:
```suggestion
As you have seen in this guide, you get distributed tracing up and running fast with the Micronaut framework without any annotations.
```

File: guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc (hunk 1 above)
Comment (timyates) [reply guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc]: Removed 👍

**Pull request merged**